pkg/controllers/clusterresourceplacement/clusterresourceplacement_controller.go

```diff
@@ -89,18 +113,77 @@ func (r *Reconciler) handleUpdate(ctx context.Context, crp *fleetv1beta1.Cluster
 
 		if err := r.Client.Create(ctx, latestPolicySnapshot); err != nil {
 			klog.ErrorS(err, "Failed to create new clusterPolicySnapshot", "clusterPolicySnapshot", policySnapshotKObj)
-			return ctrl.Result{}, controller.NewAPIServerError(false, err)
+			return nil, controller.NewAPIServerError(false, err)
 		}
 	}
+	return latestPolicySnapshot, nil
+}
 
-	// create clusterResourceSnapshot
-	// TODO
-	if _, err := r.selectResourcesForPlacement(crp); err != nil {
-		return ctrl.Result{}, err
+// TODO handle all the resources selected by placement larger than 1MB size limit of k8s objects.
+func (r *Reconciler) getOrCreateClusterResourceSnapshot(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement, resourceSnapshotSpec *fleetv1beta1.ResourceSnapshotSpec) (*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceHash, err := generateResourceHash(resourceSnapshotSpec)
+	if err != nil {
+		klog.ErrorS(err, "Failed to generate resource hash of crp", "clusterResourcePlacement", klog.KObj(crp))
+		return nil, controller.NewUnexpectedBehaviorError(err)
 	}
-	// update the status based on the latestPolicySnapshot status
-	// update the status based on the work
-	return ctrl.Result{}, nil
+	latestResourceSnapshot, latestResourceSnapshotIndex, err := r.lookupLatestResourceSnapshot(ctx, crp)
+	if err != nil {
+		return nil, err
+	}
+	latestResourceSnapshotHash := ""
```

Comment:
If you are already retrieving latest resource snapshot and its index in the lookup function, why not retrieve hashes there? It would simplify the logic a little bit.

Reply:
The lookup func is little complicated for now, so prefer to get the hash outside.

Refactored the code. PTAL to see if the readability is better.